core_lib/src/managers/backupmanager.cpp

```diff
@@ -0,0 +1,603 @@
+{
+    return Status::OK;
+}
+
+Status BackupManager::save(Object* /*o*/)
+{
+    return Status::OK;
+}
+
+const BackupElement* BackupManager::currentBackup()
+{
+    if (mUndoStack->count())
+    {
+        return static_cast<const BackupElement*>(mUndoStack->command(mUndoStack->index()-1));
+    }
+    else
+    {
+        return nullptr;
+    }
+}
+
+void BackupManager::keyAdded(int keySpacing, bool keyExisted, QString description)
+{
+    if (mLayer == NULL) { return; }
+
+    AddKeyFrameElement* element = new AddKeyFrameElement(mFrameIndex,
+                                                         mLayerId,
+                                                         keySpacing,
+                                                         keyExisted,
+                                                         description,
+                                                         editor());
+    mUndoStack->push(element);
+
+    emit updateBackup();
+}
+
+void BackupManager::keyAdded()
+{
+    if (mLayer == NULL) { return; }
+
+    AddKeyFrameElement* element = new AddKeyFrameElement(mFrameIndex,
+                                                         mLayerId,
+                                                         false,
+                                                         false,
+                                                         "",
+                                                         editor());
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+void BackupManager::importBitmap(std::map<int, KeyFrame*, std::greater<int>> canvasKeys,
+                                 std::map<int, KeyFrame*, std::less<int>> importedKeys)
+{
+    if (mLayer->type() != Layer::BITMAP) { return; }
+
+    ImportBitmapElement* element = new ImportBitmapElement(canvasKeys,
+                                                         importedKeys,
+                                                         mLayerId,
+                                                         editor());
+
+    mUndoStack->push(element);
+    emit updateBackup();
+}
+
+void BackupManager::keyRemoved()
+{
+    if (mLayer == NULL) { return; }
+
+    RemoveKeyFrameElement* element = new RemoveKeyFrameElement(mKeyframe,
+                                                               mLayerId,
+                                                               editor());
+    mUndoStack->push(element);
+    emit updateBackup();
+
+}
+
+void BackupManager::bitmap(QString description)
+{
+    if (!mBitmap) { return; }
+    AddBitmapElement* element = new AddBitmapElement(mBitmap,
+                                                     mBufferImage->clone(),
+                                                     mLayerId,
+                                                     mFrameIndex,
+                                                     description,
+                                                     editor());
+
+    new TransformElement(mKeyframe,
```

Comment:
@CandyFace I have a question about `new TransformElement`. When and who will be responsible to release this object?

Reply:
Never mind. I figured out that it has a parent object.

Reply:
Sorry for the late reply, yeah it's a child of BitmapElement, kind of obscure Qt syntax though so I can understand your question ;)

Reply:
Another question. Is there any specific reason to chain the AddBitmapElement & TransformElement together?

Reply:
Yes, because when you undo a transformation you're copying and creating a new temp bitmap image that will replace the previous. 

look at line 335 and 416 in Backuplement
```
void AddBitmapElement::undoTransform()
void AddBitmapElement::redoTransform()
```

I've added various comments about what undo does and why it requires the transform element.
It's basically for situation where when you undo/redo while having something selected and transformed. we need to know the transformation, so we can apply or discard it, therefore BitmapElement is not enough.

Reply:
Ah, ok. Thanks